src/libcore/iter.rs

```diff
@@ -2771,29 +2776,22 @@ macro_rules! range_exact_iter_impl {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<A: Int> Iterator for ::ops::Range<A> {
+#[allow(deprecated)]
+impl<A> Iterator for ::ops::Range<A> where
+    A: PartialOrd + One, for<'a> &'a A: Add<&'a A, Output = A>
+{
     type Item = A;
 
     #[inline]
     fn next(&mut self) -> Option<A> {
         if self.start < self.end {
-            let result = self.start;
-            self.start = self.start + Int::one();
-            Some(result)
+            let mut n = &self.start + &A::one();
+            mem::swap(&mut n, &mut self.start);
+            Some(n)
         } else {
             None
         }
     }
-
-    #[inline]
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        if self.start >= self.end {
-            (0, Some(0))
-        } else {
-            let length = (self.end - self.start).to_usize();
-            (length.unwrap_or(0), length)
-        }
-    }
```

Comment:
This may end up having unfortunate perf side effects, but we may be able to regain it later